sprout/appliances/tasks/__init__.py

```diff
@@ -0,0 +1,111 @@
+import hashlib
+import re
+from functools import wraps
+
+import fauxfactory
+import iso8601
+from celery import shared_task
+from django.core.cache import cache
+
+from appliances.models import Template
+from sprout import settings
+from sprout.log import create_logger
+
+LOCK_EXPIRE = 60 * 15  # 15 minutes
+
+
+def docker_vm_name(version, date):
+    return 'docker-{}-{}-{}'.format(
+        re.sub(r'[^0-9a-z]', '', version.lower()),
+        re.sub(r'[^0-9]', '', str(date)),
+        fauxfactory.gen_alphanumeric(length=4).lower())
+
+
+def parsedate(d):
+    if d is None:
+        return d
+    else:
+        return iso8601.parse_date(d)
+
+
+def singleton_task(*args, **kwargs):
+    kwargs["bind"] = True
+    wait = kwargs.pop('wait', False)
+    wait_countdown = kwargs.pop('wait_countdown', 10)
+    wait_retries = kwargs.pop('wait_retries', 30)
+
+    def f(task):
+        @wraps(task)
+        def wrapped_task(self, *args, **kwargs):
+            self.logger = create_logger(task)
+            # Create hash of all args
+            digest_base = "/".join(str(arg) for arg in args)
+            keys = sorted(kwargs.keys())
+            digest_base += "//" + "/".join("{}={}".format(key, kwargs[key]) for key in keys)
+            digest = hashlib.sha256(digest_base.encode('utf-8')).hexdigest()
+            lock_id = '{0}-lock-{1}'.format(self.name, digest)
+
+            if cache.add(lock_id, 'true', LOCK_EXPIRE):
+                try:
+                    return task(self, *args, **kwargs)
+                except Exception as e:
+                    self.logger.error(
+                        "An exception occured when executing with args: %r kwargs: %r",
+                        args, kwargs)
+                    self.logger.exception(e)
```

Comment:
This is kinda redundant, consider putting the error string ("An exception occurred ...") here, since logger.exception will automatically log the exception object repr.